explainer.md

```diff
@@ -215,6 +246,226 @@ third-party then signs either the exchanges in a package or the package as a
+validation is a call to an OS-controlled library and may make further URL
+requests that aren't under the browser's control and therefore use a separate,
+probably ephemeral, cookie jar, cache, and set of service workers.
+
+Each signature with a valid certificate chain is passed on to the next step.
+
+### Signature verification
+
+Once the certificates are validated and enough of the outer resource is received
+to parse the claimed inner headers, the client extracts the publishing URL from
+those headers and then tries to find a valid signature over the headers that is
+trusted for the publishing URL's origin. If none of the signatures are valid and
+trusted, it either
+
+1. redirects to the publishing URL as if the outer response were a 302 redirect,
+   or
+2. fails with a network error.
+
+We're not yet certain which behavior is best. The first is slightly more
+resilient to clock skew, while the second encourages intermediates to get their
+implementations right.
+
+If the client does find a valid signature, it "releases" the headers for further
+processing and starts validating
+[mi-sha256](https://tools.ietf.org/html/draft-thomson-http-mice-02#section-2)
+records into a response stream as they arrive.
+
+### Prefetching stops here
+
+At this point, the client's behavior depends on whether the outer exchange was
+requested as a [prefetch](https://w3c.github.io/resource-hints/#prefetch). To
+satisfy the [privacy-preserving prefetch](#privacy-preserving-prefetch) use
+case, prefetches have to be careful not to store the inner response in the HTTP
+cache or anywhere else that would be visible to the publishing URL's server.
+
+Prefetches can and should process any `Link: <>; rel=preload` headers they find,
+as prefetches. If those point at signed exchanges, this process repeats.
+
+### No nested signed exchanges
+
+To limit the complexity of the implementation, we're currently planning to
+disallow signed exchanges that contain either signed exchanges or redirects.
+This may change if use cases come up or if the implementation turns out to be
+simpler than expected.
+
+### Navigations and subresources redirect
+
+At this point the client redirects to the signed exchange's publishing URL, with
+the inner request and response stream "attached". (TBD how this fits into
+[Fetch](https://fetch.spec.whatwg.org/)'s terminology.) For navigations, this
+request goes through the publishing URL's Service Worker. For subresources, it
+goes through the embedder's Service Worker again. (TODO: Check that redirects
+actually work this way.)
+
+If the Service Worker fetches a matching URL from the network, either by
+returning without calling `e.respondWith()` or by calling `fetch(...)`, this
+tries to return the response stream that was attached to the redirect. However,
+if either of the following conditions is met, the fetch bypasses the attached
+exchange and continues down to the lower caches and the network:
```

Comment:
Does this extra check run only when the fetch comes from/via the Service Worker?  We also say the response is not really distinguishable between inner responses as a part of signed exchange vs regular resources from SW for now, they feel slightly inconsistent to me.

Reply:
The extra check should happen even when there's no SW or no `fetch` handler in the SW. I tend to think of a page without a SW as equivalent to a page whose SW does nothing in its `fetch` handler, so if it looks like I'm deviating from that, or if that's the wrong model for me to have, let me know. :)